core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java

```diff
@@ -0,0 +1,1214 @@
+ * copy of the License at:
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.linecorp.armeria.common;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.netty.handler.codec.http2.Http2Headers.PseudoHeaderName.hasPseudoHeaderFormat;
+import static io.netty.util.internal.MathUtil.findNextPositivePowerOfTwo;
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Arrays;
+import java.util.Date;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import javax.annotation.Nullable;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import io.netty.handler.codec.DateFormatter;
+import io.netty.util.AsciiString;
+
+/**
+ * The base container implementation of HTTP/2 headers.
+ */
+class HttpHeadersBase implements HttpHeaderGetters {
+
+    static final int DEFAULT_SIZE_HINT = 16;
+
+    /**
+     * Constant used to seed the hash code generation. Could be anything but this was borrowed from murmur3.
+     */
+    static final int HASH_CODE_SEED = 0xc2b2ae35;
+
+    private final HeaderEntry[] entries;
+    private final byte hashMask;
+
+    private final HeaderEntry head;
+    private HeaderEntry firstNonPseudo;
+
+    int size;
+    private boolean immutable;
+    private boolean endOfStream;
+
+    @Nullable
+    private HttpMethod method;
+    @Nullable
+    private HttpStatus status;
+    @Nullable
+    private MediaType contentType;
+
+    HttpHeadersBase(int sizeHint) {
+        // Enforce a bound of [2, 128] because hashMask is a byte. The max possible value of hashMask is
+        // one less than the length of this array, and we want the mask to be > 0.
+        entries = new HeaderEntry[findNextPositivePowerOfTwo(max(2, min(sizeHint, 128)))];
+        hashMask = (byte) (entries.length - 1);
+        head = firstNonPseudo = new HeaderEntry();
+    }
+
+    /**
+     * Creates an immutable shallow copy or a mutable deep copy of the specified {@link HttpHeadersBase}.
+     * <ul>
+     *   <li>If the {@code immutable} parameter is {@code true}, both the specified {@link HttpHeadersBase} and
+     *       this {@link HttpHeadersBase} become immutable.</li>
```

Comment:
Since this is a private class, it's probably ok but I was quite surprised to see the parameter become immutable

Reply:
Agreed. Let me introduce a named factory method or something similar.

Reply:
Just removed the `immutable` field and made sure it is impossible to call the setter methods. As a bonus, a user can continue building after calling `build()`:

```java
HttpHeadersBuilder b = HttpHeaders.builder();
b.add("a", "b");
HttpHeaders h1 = b.build();
b.add("c", "d");
HttpHeaders h2 = b.build();

assert h1.toString().equals("{a=b}");
assert h2.toString().equals("{a=b, c=d}");
```